src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs

```diff
@@ -0,0 +1,329 @@
+
+                var convertToHexStringLower = convertToHexStringLowerMethods.GetFirstOrDefaultMemberWithParameterTypes(byteArrayType);
+                var convertToHexStringLowerRos = rosByteType is not null ? convertToHexStringLowerMethods.GetFirstOrDefaultMemberWithParameterTypes(rosByteType) : null;
+                var convertToHexStringLowerStartLength = convertToHexStringLowerMethods.GetFirstOrDefaultMemberWithParameterTypes(byteArrayType, int32Type, int32Type);
+
+                // The following replacements are optional: Convert.ToHexStringLower is available as of .NET 9.
+                var bitConverterReplacementsToLowerBuilder = ImmutableDictionary.CreateBuilder<IMethodSymbol, IMethodSymbol>();
+                // BitConverter.ToString(data).Replace("-", "").ToLower() => Convert.ToHexStringLower(data)
+                bitConverterReplacementsToLowerBuilder.AddKeyValueIfNotNull(bitConverterToString, convertToHexStringLower);
+                // BitConverter.ToString(data, start).Replace("-", "").ToLower() => Convert.ToHexStringLower(data.AsSpan().Slice(start))
+                bitConverterReplacementsToLowerBuilder.AddKeyValueIfNotNull(bitConverterToStringStart, convertToHexStringLowerRos);
+                // BitConverter.ToString(data, start, length).Replace("-", "").ToLower() => Convert.ToHexStringLower(data, start, length)
+                bitConverterReplacementsToLowerBuilder.AddKeyValueIfNotNull(bitConverterToStringStartLength, convertToHexStringLowerStartLength);
+                var bitConverterReplacementsToLower = bitConverterReplacementsToLowerBuilder.ToImmutableDictionary();
+
+                symbols = new RequiredSymbols(stringReplaceMethods, stringToLowerMethods,
+                    stringEmptyField, bitConverterReplacements, bitConverterReplacementsToLower);
+
+                return true;
+
+                bool FilterStringReplaceMethods(IMethodSymbol stringReplaceMethod)
+                {
+                    return stringReplaceMethod.Parameters.Length >= 2 &&
+                        SymbolEqualityComparer.Default.Equals(stringReplaceMethod.Parameters[0].Type, stringType) &&
+                        SymbolEqualityComparer.Default.Equals(stringReplaceMethod.Parameters[1].Type, stringType);
+                }
+            }
+
+            /// <summary>
+            /// Attempts to get a BitConverter.ToString invocation, its appropriate replacement method and optionally a string.ToLower* invocation:
+            ///   1. Check if invocation is a string.ToLower* invocation. If so, continue the next steps with its instance.
+            ///   2. Check if invocation is a string.Replace("-", "") invocation. Abort if not. Continue the next steps with its instance.
+            ///   3. Check if invocation is a string.ToLower* invocation. If so, continue the next steps with its instance.
+            ///   4. Check if invocation is a BitConverter.ToString invocation. If not, abort.
```

Comment:
I did not understood why the same step taken at row 1 and 3, first thought it was a typo, but the implementation looks like it indeed repeats the step, this makes me feel there is a repeated logic.

It looks checks `string.Replace` first, then matches `BitConvertor.ToString()`. because this will be called for each invocation this matches twice for `BitConverter.ToString(data).Replace("-", "").ToLower()`, looks there is workaround that omit one of the matches when there `string.ToLower` involved.

I think its better to match `BitConvertor.ToString()` first, then check if it has parent invocation `string.Replace`, then the optional `string.ToLower`. This could avoid extra match and I think will be more efficient, what you think?

Reply:
This is NIT, you can keep it if you think the current solution is better.

Reply:
I agree, starting with `BitConverter.ToString()` is better to avoid duplicate work and to get rid of the duplicate diagnostics workaround.

Checking for `string.ToLower*` twice is still necessary because the following cases are both valid (and can be fixed):

1. `BitConverter.ToString(data).ToLower().Replace("-", "")`
2. `BitConverter.ToString(data).Replace("-", "").ToLower()`

I've changed the order of invocation analysis and added additional comments (with examples) to better illustrate that `string.ToLower*` can be in two places in the call chain in fe1ca9a5e133895a13de4676e27e2a94cc1808da.

Reply:
Thank you! Looks great to me